QuickOSM/core/process.py

```diff
@@ -159,18 +162,45 @@ def open_file(
                                  layer + "_colour.qml"))
 
             # Add action about OpenStreetMap
-            add_actions(new_layer, item['tags'])
+            actions.add_actions(new_layer, item['tags'])
+
+            QgsProject.instance().addMapLayer(new_layer)
 
             if final_query:
                 QgsExpressionContextUtils.setLayerVariable(
                     new_layer, 'quickosm_query', final_query)
+                actions.add_actions(new_layer, [final_layer_name], 'Layer', dialog)
 
-            QgsProject.instance().addMapLayer(new_layer)
             num_layers += 1
 
     return num_layers
 
 
+def reload_query(
+        query: str,
+        layer_name: str = 'Reloaded_query',
+        dialog: QDialog = None,
+        new_file: bool = True):
+    """ Reload a query with only the query """
+    # Getting the default overpass api and running the query
+    server = get_setting('defaultOAPI', OVERPASS_SERVERS[0]) + 'interpreter'
+
+    query = QueryPreparation(query, overpass=server)
+    final_query = query.prepare_query()
+    url = query.prepare_url()
+    connexion_overpass_api = ConnexionOAPI(url)
+    LOGGER.debug('Encoded URL: {}'.format(url))
+    osm_file = connexion_overpass_api.run()
+
+    if new_file:
+        layer_name += "_reloaded"
```

Comment:
Did you try to replace the datasource like the plugin ?

Reply:
It seems complex. The datasources for layer in memory describe only the structure, we lose all the features.

Reply:
Ok, thanks for checking !